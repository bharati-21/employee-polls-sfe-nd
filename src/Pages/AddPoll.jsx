import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { hideLoading, showLoading } from "react-redux-loading-bar";
import { useNavigate } from "react-router-dom";

import { updateCreatedQuestions, addQuestion } from "../actions";
import { _saveQuestion } from "../_DATA";

const OPTIONS_INITIAL_STATE = {
	optionOneText: "",
	optionTwoText: "",
};

const AddPoll = () => {
	const { authedUser } = useSelector((state) => state);
	const dispatch = useDispatch();
	const navigate = useNavigate();

	const [options, setOptions] = useState(OPTIONS_INITIAL_STATE);
	const [loading, setLoading] = useState(false);

	const handleOptionChange = (e) => {
		const { name, value } = e.target;
		setOptions((prevOptions) => ({
			...prevOptions,
			[name]: value,
		}));
	};

	const handlePollSubmit = async (e) => {
		e.preventDefault();
		dispatch(showLoading());
		setLoading(true);
		setOptions(OPTIONS_INITIAL_STATE);
		try {
			const response = await _saveQuestion({
				...options,
				author: authedUser,
			});
			dispatch(addQuestion(response));
			dispatch(
				updateCreatedQuestions({
					authedUser,
					questionId: response.id,
				})
			);
			dispatch(hideLoading());
			navigate("/");
		} catch (error) {
			setLoading(false);
			dispatch(hideLoading());
			console.warn(error);
			alert("Error occurred while saving poll. Please try again.");
		}
	};

	const { optionOneText, optionTwoText } = options;

	return (
		<div className="min-h-[80vh] h-full max-w-6xl mx-auto p-8 w-full gap-6 flex flex-col items-center justify-center">
			<h2 className="text-3xl">Create your own Poll</h2>
			<div className="poll-form flex flex-col gap-4 mt-2 w-full max-w-[500px]">
				<h3 className="mb-4 text-xl">Would you rather</h3>
				<form
					className="flex flex-col gap-2"
					onSubmit={handlePollSubmit}
				>
					<label className="flex flex-col gap-1">
						<span>Option one</span>
						<input
							type="text"
							className="form-control border rounded-sm border-slate-400 text-sm p-1.5"
							placeholder="Option one"
							name="optionOneText"
							required
							value={optionOneText}
							onChange={handleOptionChange}
						/>
					</label>
					<label className="flex flex-col gap-1">
						<span>Option two</span>
						<input
							type="text"
							className="form-control border rounded-sm border-slate-400 text-sm p-1.5"
							placeholder="Option two"
							required
							value={optionTwoText}
							name="optionTwoText"
							onChange={handleOptionChange}
						/>
					</label>
					<input
						type="submit"
						value="Submit Poll"
						className="btn-primary mt-4 text-base"
						disabled={loading}
					/>
				</form>
			</div>
		</div>
	);
};

export { AddPoll };
