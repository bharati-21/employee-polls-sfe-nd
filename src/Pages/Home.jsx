import { useState } from "react";
import { useSelector } from "react-redux";

import { FilteredQuestions } from "../components";
import { getFilteredQuestions, getSortedQuestionsByTimestamp } from "../utils";

const TABS = [
	{
		tabIndex: 0,
		tabName: "Unanswered",
	},
	{
		tabIndex: 1,
		tabName: "Answered",
	},
];

const Home = () => {
	const [activeTab, setActiveTab] = useState(0);
	const { questions, authedUser } = useSelector((state) => state);

	const filteredQuestions = getFilteredQuestions(
		questions,
		activeTab,
		authedUser
	);
	const sortedFilteredQuestions =
		getSortedQuestionsByTimestamp(filteredQuestions);

	const handleToggleTab = (e) => {
		const { value } = e.target;
		setActiveTab(parseInt(value));
	};

	return (
		<div className="min-h-screen p-8 max-w-6xl flex flex-col justify-start mx-auto">
			<div
				className="relative w-full p-5 py-6 h-8 flex justify-start gap-8 shadow-2xl shadow-900/20 overflow-hidden transition items-center bg-slate-800 text-gray-100 text-lg"
				role="tablist"
				aria-label="tabs"
			>
				{TABS.map(({ tabName, tabIndex }) => (
					<button
						role="tab"
						className={`inline-block hover:text-yellow-400 font-semibold ${
							activeTab === tabIndex
								? "border-b-2 border-b-yellow-500"
								: "border-b-none"
						}`}
						value={tabIndex}
						onClick={handleToggleTab}
						key={tabIndex}
					>
						{tabName}
					</button>
				))}
			</div>
			<FilteredQuestions filteredQuestions={sortedFilteredQuestions} />
		</div>
	);
};

export { Home };
