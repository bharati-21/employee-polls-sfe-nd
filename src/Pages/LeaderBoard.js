import { useSelector } from "react-redux";

import { TableBody, TableHead } from "../components";
import { getSortedUsersByQuestionsAnswers } from "../utils";

const TABLE_HEADINGS = [
	{
		id: 0,
		name: "Name",
	},
	{
		id: 1,
		name: "Answered",
	},
	{
		id: 2,
		name: "Created",
	},
];

const LeaderBoard = () => {
	const {
		users = [],
		questions = [],
		authedUser = "",
	} = useSelector((state) => state);

	const sortedUsers = getSortedUsersByQuestionsAnswers(users);
	const isDataLoading = questions.length === 0 || users.length === 0;

	return isDataLoading ? null : (
		<div className="min-h-[90vh] h-full w-full max-w-6xl mx-auto p-8 flex flex-col gap-4 justify-center">
			<h1 className="text-3xl">Leaderboard</h1>
			<div className="relative overflow-x-auto shadow-md sm:rounded-lg">
				<table className="w-full text-sm text-left text-gray-400">
					<TableHead columns={TABLE_HEADINGS} />
					<TableBody users={sortedUsers} authedUser={authedUser} />
				</table>
			</div>
		</div>
	);
};

export { LeaderBoard };
