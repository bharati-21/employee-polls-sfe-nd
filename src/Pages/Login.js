import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useLocation, useNavigate } from "react-router-dom";

import { loginUser } from "../actions/";
import { useInitData } from "../hooks/useInitData";

const Login = () => {
	const navigate = useNavigate();
	const dispatch = useDispatch();
	const location = useLocation();
	const { users } = useSelector((state) => state);

	const [user, setUser] = useState({
		userId: "",
		password: "",
	});

	const { initUsers } = useInitData();
	useEffect(() => {
		try {
			initUsers();
		} catch (error) {
			console.warn(error);
			alert("Error occured while fetching users. Please try again.");
		}
	}, [initUsers]);

	const handleSubmit = (e) => {
		e.preventDefault();

		dispatch(loginUser(user.userId));
		navigate(location?.state?.from ?? "/", { replace: true });
	};

	const handleUserSelected = (e) => {
		const id = e.target.value;
		const { password = "" } = users.find((user) => user.id === id);

		setUser({
			userId: id ?? "",
			password: password ?? "",
		});
	};

	const { userId = "", password = "" } = user;
	console.log(userId, password);

	return (
		<div className="flex h-screen items-center justify-center py-1-5 px-4 sm:px-6 lg:px-8">
			<div className="w-full max-w-md space-y-8">
				<h1 className="text-center text-4xl">Employee Polls</h1>
				<div>
					<h2 className="mt-6 text-center text-2xl font-bold tracking-tight text-gray-900">
						Sign in
					</h2>
				</div>
				<form className="mt-8 space-y-6" onSubmit={handleSubmit}>
					<label className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
						Select an option
						<select
							id="countries"
							className=" text-sm rounded-lg focus:ring-yellow-600 focus:border-yellow-600 block w-full p-2.5 bg-gray-700 border-gray-600 placeholder-gray-400 text-white"
							onChange={handleUserSelected}
							value={userId}
						>
							<option value="" disabled>
								Select a user
							</option>
							{users.map((u) => (
								<option key={u.id} value={u.id}>
									{u.name}
								</option>
							))}
						</select>
					</label>
					<div className="shadow-sm">
						<div className="text-left mb-4">
							<label>
								<span className="text-lg inline-block mb-1">
									Username
								</span>
								<input
									id="username"
									name="username"
									type="text"
									required
									className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-yellow-500 focus:outline-none focus:ring-yellow-500 sm:text-sm"
									placeholder="Username"
									disabled
									value={userId}
								/>
							</label>
						</div>
						<div className="text-left">
							<label>
								<span className="text-lg inline-block mb-1">
									Password
								</span>
								<input
									id="password"
									name="password"
									type="password"
									required
									className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-yellow-500 focus:outline-none focus:ring-yellow-500 sm:text-sm"
									placeholder="Password"
									disabled
									value={password}
								/>
							</label>
						</div>
					</div>

					<div>
						<input
							type="submit"
							disabled={!userId}
							className="btn-primary-full p-2 text-base"
							value="Sign in as Guest"
						/>
					</div>
				</form>
			</div>
		</div>
	);
};

export { Login };
