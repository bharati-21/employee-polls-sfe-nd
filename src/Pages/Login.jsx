import { useDispatch } from "react-redux";
import { useLocation, useNavigate } from "react-router-dom";

import { loginUser } from "../actions/";

const Login = () => {
	const navigate = useNavigate();
	const dispatch = useDispatch();
	const location = useLocation();

	const handleSubmit = (e) => {
		e.preventDefault();

		dispatch(loginUser("sarahedo"));
		navigate(location?.state?.from ?? "/", { replace: true });
	};
	return (
		<div className="flex h-screen items-center justify-center py-1-5 px-4 sm:px-6 lg:px-8">
			<div className="w-full max-w-md space-y-8">
				<h1 className="text-center text-4xl">Employee Polls</h1>
				<div>
					<h2 className="mt-6 text-center text-2xl font-bold tracking-tight text-gray-900">
						Sign in
					</h2>
				</div>
				<form className="mt-8 space-y-6" onSubmit={handleSubmit}>
					<div className="-space-y-px shadow-sm">
						<div className="text-left mb-4">
							<label>
								<span className="text-lg inline-block mb-1">
									Username
								</span>
								<input
									id="username"
									name="username"
									type="text"
									required
									className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm"
									placeholder="Username"
									disabled
									value="sarahedo"
								/>
							</label>
						</div>
						<div className="text-left">
							<label>
								<span className="text-lg inline-block mb-1">
									Password
								</span>
								<input
									id="password"
									name="password"
									type="password"
									required
									className="relative block w-full appearance-none rounded border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500 sm:text-sm"
									placeholder="Password"
									disabled
									value="password123"
								/>
							</label>
						</div>
					</div>

					<div>
						<input
							type="submit"
							className="group relative flex w-full justify-center rounded border border-transparent bg-indigo-600 py-2 px-4 font-medium text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2 cursor-pointer text-md"
							value="Sign in as Guest"
						/>
					</div>
				</form>
			</div>
		</div>
	);
};

export { Login };
