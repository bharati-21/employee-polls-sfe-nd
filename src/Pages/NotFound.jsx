import React from "react";
import { Link } from "react-router-dom";

const NotFound = () => {
	return (
		<div className="h-screen flex flex-col justify-center items-center gap-8 p-8">
			<h1 className="text-4xl">404! Page Not Found!</h1>
			<Link to="/" className="btn-primary">
				Go Back Home
			</Link>
		</div>
	);
};

export { NotFound };
