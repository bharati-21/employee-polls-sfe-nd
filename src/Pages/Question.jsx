import React, { useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";

import { removeVoteQuestion, voteQuestion, updateAnswered } from "../actions/";
import { _saveQuestionAnswer } from "../_DATA";

const Question = () => {
	const { questionId } = useParams();
	const {
		questions = [],
		users = [],
		authedUser,
	} = useSelector((state) => state);
	const dispatch = useDispatch();
	const navigate = useNavigate();

	const [loading, setLoading] = useState(false);

	const question = useMemo(() => {
		return questions.find((question) => question.id === questionId) ?? {};
	}, [questions, questionId]);
	const {
		author = "",
		optionOne: { text: textOne = "", votes: votesOne = [] } = {},
		optionTwo: { text: textTwo = "", votes: votesTwo = [] } = {},
	} = question;

	const user = useMemo(() => {
		return users.find((user) => user.id === author) ?? {};
	}, [users, author]);
	const { avatarURL = "" } = user;

	const isValidQuestion = Object.keys(question).length !== 0;

	useEffect(() => {
		// invalid question id
		if (questions.length && !isValidQuestion) {
			navigate("/error", { replace: true });
		}
	}, [navigate, question, questions, isValidQuestion]);

	const getQuestionOption = (option) => {
		switch (option) {
			case "1":
				return "optionOne";
			case "2":
				return "optionTwo";
			default:
				return 0;
		}
	};

	const isAuthorAnswered = (option) => {
		switch (option) {
			case 1:
				if (votesOne.includes(authedUser)) {
					return true;
				}
				return false;
			case 2:
				if (votesTwo.includes(authedUser)) {
					return true;
				}
				return false;
			default:
				return false;
		}
	};

	const handleVoteQuestion = async (e) => {
		const { value } = e.target;
		const option = getQuestionOption(value);

		setLoading(true);

		// optimistic update
		dispatch(voteQuestion({ questionId, option, authedUser }));
		try {
			await _saveQuestionAnswer({
				authedUser,
				qid: questionId,
				answer: option,
			});
			dispatch(
				updateAnswered({
					authedUser,
					questionId,
				})
			);
			setLoading(false);
		} catch (error) {
			alert("Some error occurred while voting. Please try again.");
			dispatch(
				removeVoteQuestion({
					questionId,
					option,
					authedUser,
				})
			);
			setLoading(false);
		}
	};

	const isLoading = !isValidQuestion || !questions.length || !users.length;
	const btnDisabled = loading || isAuthorAnswered(1) || isAuthorAnswered(2);

	return isLoading ? null : (
		<div className="min-h-[90vh] h-full mx-auto p-8 max-w-6xl w-full justify-center items-center flex flex-col">
			<img
				className="inline-block h-[150px] w-[150px] rounded-full ring-2 object-cover ring-white"
				src={avatarURL}
				alt={`${authedUser} avatar`}
			/>
			<div className="question-wrapper p-1 flex flex-wrap gap-8 justify-center">
				<h1 className="text-xl mt-5">Would you rather</h1>
				<div className="question-options-container flex flex-wrap gap-2 w-full">
					<div className="option-one flex flex-col flex-1 justify-between">
						<span className="border-gray-300 border-2 border-b-0 py-1.5 flex-1">
							{textOne}
						</span>
						<button
							className={
								isAuthorAnswered(1)
									? "btn-voted"
									: "btn-primary"
							}
							disabled={btnDisabled}
							value={1}
							onClick={handleVoteQuestion}
						>
							Vote
						</button>
					</div>
					<div className="option-two flex flex-col flex-1 justify-between">
						<span className="border-gray-300 border-2 border-b-0 py-1.5 flex-1">
							{textTwo}
						</span>
						<button
							className={
								isAuthorAnswered(2)
									? "btn-voted"
									: "btn-primary"
							}
							disabled={btnDisabled}
							value={2}
							onClick={handleVoteQuestion}
						>
							Vote
						</button>
					</div>
				</div>
			</div>
		</div>
	);
};

export { Question };
