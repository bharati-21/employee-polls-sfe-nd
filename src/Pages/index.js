export { Home } from "./Home";
export { LeaderBoard } from "./LeaderBoard";
export { Login } from "./Login";
export { NotFound } from "./NotFound";
export { Question } from "./Question";
export { AddPoll } from "./AddPoll";
