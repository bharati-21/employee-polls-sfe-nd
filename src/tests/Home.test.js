import { render } from "@testing-library/react";
import * as React from "react";
import { Home } from "../Pages";
import StoreProvider from "../components/StoreProvider";
import { BrowserRouter } from "react-router-dom";

describe("Home", () => {
	it("will match snapshot", () => {
		const view = render(
			<StoreProvider>
				<BrowserRouter>
					<Home />
				</BrowserRouter>
			</StoreProvider>
		);
		expect(view).toMatchSnapshot();
	});
});
