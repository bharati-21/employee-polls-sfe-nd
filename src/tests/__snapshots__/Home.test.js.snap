// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="min-h-screen p-8 max-w-6xl flex flex-col justify-start mx-auto"
      >
        <div
          aria-label="tabs"
          class="relative w-full p-5 py-6 h-8 flex justify-start gap-8 shadow-2xl shadow-900/20 overflow-hidden transition items-center bg-slate-800 text-gray-100 text-lg"
          role="tablist"
        >
          <button
            class="inline-block hover:text-yellow-400 font-semibold border-b-2 border-b-yellow-500"
            role="tab"
            value="0"
          >
            Unanswered
          </button>
          <button
            class="inline-block hover:text-yellow-400 font-semibold border-b-none"
            role="tab"
            value="1"
          >
            Answered
          </button>
        </div>
        <div
          class="bg-slate-200 flex flex-wrap justify-center gap-8 px-8 py-4"
        >
          <h1
            class="text-2xl"
          >
            You're all caught up!
          </h1>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="min-h-screen p-8 max-w-6xl flex flex-col justify-start mx-auto"
    >
      <div
        aria-label="tabs"
        class="relative w-full p-5 py-6 h-8 flex justify-start gap-8 shadow-2xl shadow-900/20 overflow-hidden transition items-center bg-slate-800 text-gray-100 text-lg"
        role="tablist"
      >
        <button
          class="inline-block hover:text-yellow-400 font-semibold border-b-2 border-b-yellow-500"
          role="tab"
          value="0"
        >
          Unanswered
        </button>
        <button
          class="inline-block hover:text-yellow-400 font-semibold border-b-none"
          role="tab"
          value="1"
        >
          Answered
        </button>
      </div>
      <div
        class="bg-slate-200 flex flex-wrap justify-center gap-8 px-8 py-4"
      >
        <h1
          class="text-2xl"
        >
          You're all caught up!
        </h1>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
